### PR TITLE
#755 FIX unit component duplication BUG

### PR DESCRIPTION
**Pull request by nimshi89**
#755 Bug was noticed during the work on #703.

### CODE REVIEW

**Review by NickPapONS — COMMENTED** [review]
Looks good to me! All I would do is tell sonarcloud that the "Existing Dimension Component" used in the new unit test is not an issue so the code smell disappears.

File: src/csvcubed/writers/helpers/qbwriter/dsdtordfmodelshelper.py [hunk 1]
```diff
@@ -301,17 +301,29 @@ def _get_qb_component_specs_for_col(
...
-            return [self._get_qb_units_column_specification(column_name_uri_safe)]
+            return self._get_qb_units_column_specification(column_name_uri_safe)
         elif isinstance(component, QbMultiMeasureDimension):
             return self._get_qb_measure_dimension_specifications(component)
         elif isinstance(component, QbObservationValue):
             return self._get_qb_obs_val_specifications(component)
         else:
             raise TypeError(f"Unhandled component type {type(component)}")
 
+    _units_component_already_defined: bool = False
```
Comment (robons) [inline src/csvcubed/writers/helpers/qbwriter/dsdtordfmodelshelper.py:312]: Could you move this up to the top of the class where the other attributes are defined. Can you also declare that this isn't an attribute which should appear in the `__init__` function? The dev using this class shouldn't be allowed to configure this themselves.

File: src/csvcubed/writers/helpers/qbwriter/dsdtordfmodelshelper.py [hunk 2]
```diff
@@ -301,17 +301,29 @@ def _get_qb_component_specs_for_col(
...
+    If it has, don't define it again.
+    """
+
     def _get_qb_units_column_specification(
         self, column_name_uri_safe: str
-    ) -> rdf.qb.AttributeComponentSpecification:
+    ) -> List[rdf.qb.AttributeComponentSpecification]:
+        if self._units_component_already_defined:
+            # Don't define a second units component, the first one will work just fine.
+            return []
```
Comment (robons) [inline src/csvcubed/writers/helpers/qbwriter/dsdtordfmodelshelper.py:323]: Can you add some debug logging here so we can track that we're avoiding serialising a second units component?

File: tests/unit/writers/qbwriter/test_dsdtordfmodelshelper.py [hunk 3]
```diff
@@ -577,5 +578,69 @@ def test_qb_order_of_components():
...
+                QbObservationValue(
+                    NewQbMeasure("Some Other Measure"), NewQbUnit("Some Other Unit")
+                ),
+            ),
+        ],
+    )
+
+    dsd_helper = DsdToRdfModelsHelper(cube, UriHelper(cube))
+    dataset = dsd_helper._generate_qb_dataset_dsd_definitions()
+    graph = dataset.to_graph(Graph())
```
Comment (robons) [inline tests/unit/writers/qbwriter/test_dsdtordfmodelshelper.py:624]: We don't need this graph or the assertion do we?

File: src/csvcubed/writers/helpers/qbwriter/dsdtordfmodelshelper.py (hunk 1 above)
Comment (nimshi89) [reply src/csvcubed/writers/helpers/qbwriter/dsdtordfmodelshelper.py:312]: sorted

File: src/csvcubed/writers/helpers/qbwriter/dsdtordfmodelshelper.py (hunk 2 above)
Comment (nimshi89) [reply src/csvcubed/writers/helpers/qbwriter/dsdtordfmodelshelper.py:328]:
```    def _get_qb_units_column_specification(
        self, column_name_uri_safe: str
    ) -> List[rdf.qb.AttributeComponentSpecification]:
        if self._units_component_already_defined:
            _logger.debug(
                "The component type has already been serialised, %s.",
                column_name_uri_safe,
            )
            # Don't define a second units component, the first one will work just fine.

            return []```

is this the right place for it?

File: tests/unit/writers/qbwriter/test_dsdtordfmodelshelper.py (hunk 3 above)
Comment (nimshi89) [reply tests/unit/writers/qbwriter/test_dsdtordfmodelshelper.py]: sorted

File: src/csvcubed/writers/helpers/qbwriter/dsdtordfmodelshelper.py [hunk 4]
```diff
@@ -301,17 +302,33 @@ def _get_qb_component_specs_for_col(
...
-            return [self._get_qb_units_column_specification(column_name_uri_safe)]
+            return self._get_qb_units_column_specification(column_name_uri_safe)
         elif isinstance(component, QbMultiMeasureDimension):
             return self._get_qb_measure_dimension_specifications(component)
         elif isinstance(component, QbObservationValue):
             return self._get_qb_obs_val_specifications(component)
         else:
             raise TypeError(f"Unhandled component type {type(component)}")
 
+    """
```
Comment (robons) [inline src/csvcubed/writers/helpers/qbwriter/dsdtordfmodelshelper.py:313]: This comment should be against the field we added (which you've now moved up).

File: src/csvcubed/writers/helpers/qbwriter/dsdtordfmodelshelper.py (hunk 2 above)
Comment (robons) [reply src/csvcubed/writers/helpers/qbwriter/dsdtordfmodelshelper.py:328]: that's definitely the right place, but could it say something like `Units component already generated. Not generating a second one.`

Comment (nimshi89) [reply src/csvcubed/writers/helpers/qbwriter/dsdtordfmodelshelper.py:328]: Yeah cool I'll sort it out

File: src/csvcubed/writers/helpers/qbwriter/dsdtordfmodelshelper.py (hunk 4 above)
Comment (nimshi89) [reply src/csvcubed/writers/helpers/qbwriter/dsdtordfmodelshelper.py:313]: moved it

**Review by robons — APPROVED** [review]

**Pull request merged**